src/main/java/htsjdk/samtools/util/GZIIndex.java

```diff
@@ -0,0 +1,365 @@
+                .findFirst().orElse(FIRST_MAPPING);
+
+        // convert to an offset within the block
+        final int blockOffset = Math.toIntExact(indexEntry.getUncompressedOffset() - uncompressedOffset);
+
+        return BlockCompressedFilePointerUtil.makeFilePointer(indexEntry.getCompressedOffset(), blockOffset);
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        if (obj == null || !(obj instanceof GZIIndex)) {
+            return false;
+        }
+        return this.entries.equals(((GZIIndex) obj).entries);
+    }
+
+    @Override
+    public int hashCode() {
+        return entries.hashCode();
+    }
+
+    @Override
+    public String toString() {
+        return "GZIIndex:" + StringUtil.join(", ", entries);
+    }
+
+    /**
+     * Writes this index into the requested path.
+     *
+     * @param output the output file.
+     *
+     * @throws IOException if an I/O error occurs.
+     */
+    public void writeIndex(final Path output) throws IOException {
+        if (output == null) {
+            throw new IllegalArgumentException("null output path");
+        }
+
+        // the first entry is never written - 0, 0
+        final int numberOfBlocksToWrite = getNumberOfBlocks() - 1;
+
+        final ByteBuffer buffer = allocateBuffer(numberOfBlocksToWrite, true);
+        // put the number of entries
+        buffer.putLong(Integer.toUnsignedLong(numberOfBlocksToWrite));
+
+        // except the first, iterate over all the entries
+        for (final IndexEntry entry : entries.subList(1, entries.size())) {
+            // implementation of entry ensures that the offsets are no negative
+            buffer.putLong(entry.getCompressedOffset());
+            buffer.putLong(entry.getUncompressedOffset());
+        }
+
+        // write into the output
+        Files.write(output, buffer.array());
+    }
+
+    /**
+     * Loads the index from the provided file.
+     *
+     * @param indexPath the path for the index to load.
+     *
+     * @return loaded index.
+     *
+     * @throws IOException if an I/O error occurs.
+     */
+    public static final GZIIndex loadIndex(final Path indexPath) throws IOException {
+        if (indexPath == null) {
+            throw new IllegalArgumentException("null input path");
+        }
+        // allocate a buffer for re-use for read each byte
+        ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES);
+        buffer.order(ByteOrder.LITTLE_ENDIAN);
+
+        try (final ByteChannel channel = Files.newByteChannel(indexPath)) {
+            if (Long.BYTES != channel.read(buffer)) {
+                throw getCorruptedIndexException(indexPath, "less than " + Long.BYTES+ "bytes");
+            }
+            buffer.flip();
+
+            final int numberOfEntries;
+            try {
+                numberOfEntries = Math.toIntExact(buffer.getLong());
+            } catch (ArithmeticException e) {
+                buffer.flip();
```

Comment:
why bother flipping the buffer here?

Reply:
To show in the error message the value of the number of entries field in the index file, because the maximum that java will accept is Integer.MAX_VALUE. I am re-using the buffer here instead of storing the long value, but I can change that....